### DISCUSSION TITLE
Features behind artifactory

### DISCUSSION THREAD

**Discussion by tclarke** [devcontainers/community#96] [Q&A]
I'm trying to use devcontainers on a machine without Internet access. All access is through a local artifactory. I can use this to specify a feature I want to use, for example:
```
"features": {
  "arti.myorg.com/docker-group/ghcr.io/bdsoha/devcontainers/custom-root-ca:1": { "source": "..." }
}
```
But I still get an error trying to grab "ghcr.io/devcontainers/features/common-utils". Is there a way to generically specify a different server for dependent features? I tried cloning the common-utils locally and adding that directory to the features map but it still tries to grab the ghcr.io version.

**Comment by joshspicer** [comment]
The easiest way would be to build your dev container prior to losing internet access.  There is an expectation that you will have internet access when building a dev container.  Lots of Features fetch tools, etc dynamically from the internet on installation.  The CLI will also attempt several network calls for various reasons.  I recently took a flight and (using the dev containers VS Code extension) made sure I launched the dev container before taking off.

Other options include:

- remove the dependency on the remote Feature from `custom-root-ca`'s `devcontainer-feature.json` (I don't have access to your source, but sounds like its a dependency?)
- clone github.com/devcontainers/features and use the dev container CLI to re-publish those Features to your local artifactory (changing the references in your other Features).  
- Copy the relevant bits of "common-utils" into a ["local Feature"](https://containers.dev/implementors/features-distribution/#addendum-locally-referenced) that you can check in beside your `devcontainer.json`.

Again, note that a lot of Features have runtime dependencies on the internet, so you'd likely need to heavily modify any Features that need to access the network.

If you have no references to the Feature and think this is a bug, if you could provide us a minimal repro to test against that would be very helpful

**Reply by tclarke** [reply]
This is on a computer in a classified network so there's 0 change of internet access. It would be nice to have a config file or similar to blanket replace certain registries with others. I.E. replace gcr.io with artifactory.local/artifactory/gcr_io/
